File: core/scope_transactions/domain/responses/account.go
package responses

import (
	"financo/lib/color"
	"financo/lib/currency"
	"financo/lib/icon"
	"financo/lib/nullable"
	"financo/models/account"
	"time"
)

type Account struct {
	ID          int64                    `json:"id"`
	ParentID    nullable.Type[int64]     `json:"parentId"`
	Kind        account.Kind             `json:"kind"`
	Currency    currency.Type            `json:"currency"`
	Name        string                   `json:"name"`
	Description nullable.Type[string]    `json:"description"`
	Color       color.Type               `json:"color"`
	Icon        icon.Type                `json:"icon"`
	Capital     int64                    `json:"capital"`
	Balance     int64                    `json:"balance"`
	Main        bool                     `json:"main"`
	ArchivedAt  nullable.Type[time.Time] `json:"archivedAt"`
	DeletedAt   nullable.Type[time.Time] `json:"deletedAt"`
	CreatedAt   time.Time                `json:"createdAt"`
	UpdatedAt   time.Time                `json:"updatedAt"`
}

func RecordToAccount(r account.Record) Account {
	return Account{
		ID:          r.ID,
		ParentID:    r.ParentID,
		Kind:        r.Kind,
		Currency:    r.Currency,
		Name:        r.Name,
		Description: r.Description,
		Color:       r.Color,
		Icon:        r.Icon,
		Capital:     r.Capital,
		Balance:     r.DynamicData.Balance,
		Main:        r.DynamicData.Main,
		ArchivedAt:  r.ArchivedAt,
		DeletedAt:   r.DeletedAt,
		CreatedAt:   r.CreatedAt,
		UpdatedAt:   r.UpdatedAt,
	}
}

File: webapp/app/modules/shared/types/account.ts
import { Currency } from "./currency";
import { Icon } from "./icon";

export {
  isCapital,
  isCategory,
  isCredit,
  isDebt,
  isExpense,
  isIncome,
  isPassive,
  isSavings
};

export type {
  Account,
  AccountKind,
  AccountPreview,
  CategoryKind,
  SystemKind as Kind,
  Kinds,
  PureCategoryKind,
  SystemKind
};

export const KINDS = {
  history: "history",
  capital: "capital",
  savings: "savings",
  debt: "debt",
  credit: "credit",
  income: "income",
  expense: "expense"
} as const;

type Kinds = typeof KINDS;

type SystemKind = Kinds["history"] |
  Kinds["capital"] |
  Kinds["savings"] |
  Kinds["debt"] |
  Kinds["credit"] |
  Kinds["income"] |
  Kinds["expense"];

type AccountKind = Kinds["capital"] |
  Kinds["savings"] |
  Kinds["debt"] |
  Kinds["credit"];

type CategoryKind = Kinds["debt"] |
  Kinds["credit"] |
  Kinds["income"] |
  Kinds["expense"];

type PureCategoryKind = Kinds["income"] | Kinds["expense"];

interface History {
  at: string | null | undefined
  balance: number | null | undefined
}

interface AdditionalData {
  main: boolean
  balance: number
  history: History
  transactions: number
}

interface Account {
  id: number
  kind: AccountKind
  currency: Currency
  name: string
  description?: string | null
  icon: Icon
  color: string
  capital: number
  additionalData: AdditionalData
  archivedAt: string | null | undefined
  deletedAt: string | null | undefined
  createdAt: string
  updatedAt: string
}

interface AccountPreview {
  id: number
  parentId: number | null
  kind: AccountKind
  currency: Currency
  name: string
  description?: string | null
  color: string
  icon: Icon
  capital: number
  balance: number
  main: boolean
  archivedAt: string | null
  deletedAt: string | null
  createdAt: string
  updatedAt: string
}

function isCapital(kind: SystemKind) {
  return kind === KINDS.capital
}

function isSavings(kind: SystemKind) {
  return kind === KINDS.savings
}

function isPassive(kind: SystemKind) {
  return kind === KINDS.debt || kind === KINDS.credit
}

function isDebt(kind: SystemKind) {
  return kind === KINDS.debt
}

function isCredit(kind: SystemKind) {
  return kind === KINDS.credit
}

function isIncome(kind: SystemKind) {
  return kind === KINDS.income
}

function isExpense(kind: SystemKind) {
  return kind === KINDS.expense
}

function isCategory(kind: SystemKind) {
  return kind === KINDS.expense || kind === KINDS.income
}
